### DISCUSSION TITLE
use external dhcp without proxy

### DISCUSSION THREAD

**Discussion by stiebel** [ltsp/ltsp#307] [Q&A]
Hello community,
I used to work with the old LTSP for roughly 10 years at a school parallel to an existing windows network.
The workstations are with windows-disks and the ltsp was booted via PXE, standard boot option was networkboot.
I used ipfire as a DHCP Server providing the tftp-boot information.

But now we have a change: the Windows administrator needs the PXE-Boot for remote admin for windows.
So I can not Use PXE-boot any more and for this reason the ltsp server must not provide any dhcp services any more,
even no dnsmasq dhcp-proxy. 
I also have to switch of my ipfire dhcp-server. The full dhcp-system will be done by windows only.

After long negotiations I got the deal to get USB-Stick booting available and as first boot option.
So without a usb-stick the Workstations looks for administration with PXE, if there is none, it starts from disk and the Windows Admin is happy.
With an attached USB Stick, it boots from there.
I used IPXE.USB to make a USB Stick being my linux-boot device.
This seems to work.

But now I need a configuration for the ltsp server to have a fixed IP and to stop providing any DHCP service.
And I need the IPXE to have the information to chainload from this fixed IP.

I think the IPXE will work "make"-ing a new ipxe.usb with additional parameters.
(like: get dhcp address, ignore provided nextboot, use fixed tftpboot)

But I did not use dnsmasq with my old LTSP and I do not now how to set this up.

I think I could work it out, but if someone did that before, I would appreciate the shortcut.
Greetings, Roland

PS: please no discussions about the sens of using windows network services. I have no choice!

**Comment by alkisg** [comment]
The easiest solution would be to build a custom ipxe, either locally or online at https://rom-o-matic.dev/ with a script similar to the following (untested):

```shell
#!ipxe

dhcp net0
set next-server 10.161.254.11
chain /ltsp/ltsp.ipxe
```

**Comment by stiebel** [comment]
Hi Alkisg, thanks for answering !
That is really great work you have done to LTSP, thanks a lot for that. It has become much easier than last time I built a server with it.

How would I set up the LTSP server to prevent it from beeing a dhcp server/proxy?
Just with the "ltsp dnsmasq proxy=0 dchp=0" line?
How do I set the IP-Address when using Ubuntu 20 without a graphical desktop?

Greetings, Roland

**Comment by stiebel** [comment]
just tried it ...
It works like I posted.
The three lines from Alkisg worked as well. Thanks again.
Roland

**Comment by stiebel** [comment]
Hi Alkisg and others,
I posted the other week that it worked.
Well, kind of. I tested it in a virtual environment and that was fine... 
But yesterday I tested it in real world and it worked like this:
ipxe booted from USB -> worked
ipxe get dhcp -> worked
ipxe overrides next-server -> worked
the ltsp pxe-menu appears
the ltsp img gets loeaded
the initrd get loaded
the system starts as expected, but
during the boot sequence the systems again sends a dhcp request and gets an ip,
but I assume it again gets the wrong tftp/nfs/nbd -server ip, 
because just after the successfully finished dhcp negotiation 
I get endless read error messages.

I tried creating an etc/ltsp/ltsp.conf file with a server="10.3.10.12" line in the [common] section,
but this did not change anything. (and yes, I placed the ltsp initrd command...)

Any suggestions?
Thanks, Roland

**Comment by alkisg** [comment]
Idea 1: try with `SERVER=ip`, not lowercase `server=ip`.

Idea 2: put KERNEL_PARAMETERS="break=mount" in ltsp.conf under [common], then run `ltsp ipxe`, then reboot the client; it should stop right before the mounting step in the initramfs; type `cat /proc/cmdline` there, and check if nfsroot=${ip} has the correct IP there or not.

Idea 3: if the above don't reveal anything, upload a video of the boot process.

**Comment by stiebel** [comment]
ok, we are getting closer.
 cat /proc/cmdline tells me : nfsroot=10.1.1.1... 
which is the wrong server.
here is my ipxe.conf file:

```shell
#!ipxe

dhcp net0
set next-server 10.3.10.12
chain /ltsp/ltsp.ipxe
```

and here is my /etc/ltsp/ltsp.conf

```shell
$ more ltsp.conf 
# /bin/sh -n
# LTSP configuration file
# Documentation=man:ltsp.conf(5)

# The special [server] section is evaluated only by the ltsp server
[server]
# Enable NAT on dual NIC servers
# NAT=1
# Provide a full menu name for x86_32.img when `ltsp ipxe` runs
# IPXE_X86_32_IMG="Debian Buster"

# The special [common] section is evaluated by both the server and ltsp clients
[common]
SERVER="10.3.10.12"
# Specify an alternative TFTP_DIR
# TFTP_DIR=/var/lib/tftpboot

KERNEL_PARAMETERS="break=mount"

# In the special [clients] section, parameters for all clients can be defined.
# Most ltsp.conf parameters should be placed here.
[clients]
# Specify an /etc/fstab line for NFS home; note this is insecure
# FSTAB_HOME="server:/home /home nfs defaults,nolock 0 0"

# MAC address, IP address, or hostname sections can be used to apply settings
# to specific clients.
[61:6c:6b:69:73:67]
# HOSTNAME=pc01
# Include parameters from another section, defined below
# INCLUDE=crt_monitor

# Shell "case" expressions can be used in MAC, IP, or hostname sections.
# This matches all Raspberry Pi MAC addresses.
[b8:27:eb:*|dc:a6:32:*]
# FSTAB_BOOT="/dev/mmcblk0p1  /boot  vfat  defaults  0  2"

# If you have proper DNS, you can use hostname sections
[administrator-pc]
# Only allow the administrator to log in to this client
# PWMERGE_SUR="administrator"

# You can also group parameters into named sections and INCLUDE= them elsewhere
[crt_monitor]
# Force EDID and resolution to 1024x768 for clients with old CRT monitors
X_HORIZSYNC="28.0-87.0"
X_VERTREFRESH="43.0-87.0"
X_MODES='"1024x768" "800x600" "640x480"
```

Where do I put the right nfsroot?`

**Comment by stiebel** [comment]
thanks for editing and setting the right code format ;)

**Comment by alkisg** [comment]
And what's your /srv/tftp/ltsp/ltsp.ipxe?

**Comment by stiebel** [comment]
got it !!!
I found a workaround:
Thanks alkisq for asking for /srv/tftp/ltsp/ltsp.ipxe
this is actually generated by /usr/share/ltsp/server/ipxe/ltsp.ipxe
and in there I hardcoded : 
```
set next-server 10.3.10.12
set srv 10.3.10.12
```
Whatever goes wrong in passing this information from somewhere else doesn't matter anymore.
I am sure there is a better an cleaner way, but for me I am fine.
Thanks a lot!
Roland

**Comment by stiebel** [comment]
Hi, 
I managed to log in with ltsp and this was a great step forward.
Now I run into two new problems:

1) the standard dhcp server provides not only ip adresses, but many more (so problems above)
and again I have to change the default behavior for the ltsp clients:
I have to change the netmask, gateway and the dns entry to my "linux pass by server"
as the standard windows gateway/proxy is bound to windows active directory users.

I have a second gateway up und running, but I have to tell the clients to use it.
At the ltsp-server itself I have a fixed ip adress with fixed gateway/dns.

2) The ltsp-server itself is a virtualbox machine.
When starting the fatclients they try to start virtualbox programms at startup. They slow down the boot process.
How do I fix this?
(at my last project I had the ltsp-server running as a KVM client, but there I had the chroot version of LTSP4
and so there was no interference between the server settings and the fatclient settings.
Would I run into similar promlems when using KVM again or would this fix all my problems?
I'd rather stick to virtualbox but I don't have to.

Thanks again for your help!
Roland

**Comment by alkisg** [comment]
1. You can use something like this in ltsp.conf. The 4 values are ip:server:gateway:netmask:

```shell
[clients]
KERNEL_PARAMETERS="ip=\${ip}:192.168.67.1:192.168.1.1:255.255.255.0"
DNS_SERVER="192.168.67.1 8.8.8.8"
```

For `KERNEL_PARAMETERS` to take effect, you need `ltsp ipxe` (the result goes in /srv/tftp/ltsp/ltsp.ipxe), while for `DNS_SERVER` to take effect, you need `ltsp initrd`.

2. Do you mean that the vbox modules are loaded? These should take less than 1 sec and so little RAM that it shouldn't be annoying... If you want to blacklist them, you can do something like `POST_INIT_BLACKLIST_VBOX="echo blacklist vboxsomething >> /etc/modprobe.d/blacklist.conf"`, i.e. standard blacklisting techniques.

**Comment by stiebel** [comment]
Hi Alkisg,
thanks again!
I just tried 1. 
the KERNEL_PARAMETERS settings sound very promising,
but 
```
KERNEL_PARAMETERS="ip=\${ip}:10.3.10.12:10.3.10.11:255.255.0.0"
DNS_SERVER="10.3.10.11 8.8.8.8"
```
in the etc/ltsp.conf did not work for me.
(yes I run sudo ltsp ipxe)
How could I check if these parameters find their way to the clients?

**Comment by alkisg** [comment]
Just open /srv/tftp/ltsp/ltsp.ipxe and see if that ip=... line is there.

To check from the client, you'd run `cat /proc/cmdline`, but it's best to first check ltsp.ipxe with an editor.
